src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -178,20 +301,499 @@ func (s *commitLogSource) ReadData(
+	nsID ident.ID,
+	shard uint32,
+	blockStart time.Time,
+	metadataOnly bool,
+	shardResult result.ShardResult,
+	allSeriesSoFar *result.Map,
+	blockSize time.Duration,
+	snapshotFiles fs.FileSetFilesSlice,
+	mostRecentCompleteSnapshot snapshotMetadata,
+) (result.ShardResult, error) {
+	var (
+		bOpts      = s.opts.ResultOptions()
+		blOpts     = bOpts.DatabaseBlockOptions()
+		blocksPool = blOpts.DatabaseBlockPool()
+		bytesPool  = blOpts.BytesPool()
+		fsOpts     = s.opts.CommitLogOptions().FilesystemOptions()
+		idPool     = s.opts.CommitLogOptions().IdentifierPool()
+	)
+
+	// Bootstrap the snapshot file
+	reader, err := s.newReaderFn(bytesPool, fsOpts)
+	if err != nil {
+		return shardResult, err
+	}
+
+	err = reader.Open(fs.DataReaderOpenOptions{
+		Identifier: fs.FileSetFileIdentifier{
+			Namespace:   nsID,
+			BlockStart:  blockStart,
+			Shard:       shard,
+			VolumeIndex: mostRecentCompleteSnapshot.ID.VolumeIndex,
+		},
+		FileSetType: persist.FileSetSnapshotType,
+	})
+	if err != nil {
+		return shardResult, err
+	}
+
+	s.log.Infof(
+		"Reading snapshot for shard: %d and blockStart: %d and volume: %d",
+		shard, blockStart.Unix(), mostRecentCompleteSnapshot.ID.VolumeIndex)
+	for {
+		var (
+			id               ident.ID
+			tagsIter         ident.TagIterator
+			data             checked.Bytes
+			expectedChecksum uint32
+		)
+		if metadataOnly {
+			id, tagsIter, _, _, err = reader.ReadMetadata()
+		} else {
+			id, tagsIter, data, expectedChecksum, err = reader.Read()
+		}
+		if err != nil && err != io.EOF {
+			return shardResult, err
+		}
+
+		if err == io.EOF {
+			break
+		}
+
+		var tags ident.Tags
+		if tagsIter.Remaining() > 0 {
+			tags, err = convert.TagsFromTagsIter(id, tagsIter, idPool)
+			if err != nil {
+				return shardResult, fmt.Errorf("unable to decode tags: %v", err)
+			}
+		}
+		tagsIter.Close()
+
+		dbBlock := blocksPool.Get()
+		dbBlock.Reset(blockStart, blockSize, ts.NewSegment(data, nil, ts.FinalizeHead))
+
+		if !metadataOnly {
+			// Resetting the block will trigger a checksum calculation, so use that instead
+			// of calculating it twice.
+			checksum, err := dbBlock.Checksum()
+			if err != nil {
+				return shardResult, err
+			}
+
+			if checksum != expectedChecksum {
+				return shardResult, fmt.Errorf("checksum for series: %s was %d but expected %d", id, checksum, expectedChecksum)
+			}
+		}
+
+		if allSeriesSoFar != nil {
+			existing, ok := allSeriesSoFar.Get(id)
+			if ok {
+				// If we've already bootstrapped this series for a different block, we don't need
+				// another copy of the IDs and tags.
+				id.Finalize()
+				tags.Finalize()
+				id = existing.ID
+				tags = existing.Tags
+			}
+		}
+
+		if shardResult == nil {
+			// Delay initialization so we can estimate size.
+			shardResult = result.NewShardResult(reader.Entries(), s.opts.ResultOptions())
+		}
+
+		// Mark the ID and Tags as no finalize because we're sharing them across blocks.
+		// TODO: Not sure this is necessary...
```

Comment:
Hey so this is not necessary I believe, since the shardResult keeps series metadata just once per collection of blocks.

Reply:
I added this to m3x here: